pkg/scheduler/eventhandlers/reservation_handler.go

```diff
@@ -0,0 +1,150 @@
+		if err == nil {
+			klog.V(4).InfoS("add reservation, enqueue reserve pod to scheduling queue",
+				"reservation", reservation.GetReservationNamespacedName(r))
+		} else {
+			klog.Errorf("failed to add reserve pod into scheduling queue, reservation %v, err: %v",
+				reservation.GetReservationNamespacedName(r), err)
+		}
+	}
+}
+
+func updateReservation(sched *scheduler.Scheduler, oldR, newR *schedulingv1alpha1.Reservation) {
+	if oldR == nil || newR == nil {
+		klog.Errorf("failed to update reservation for invalid reservation old %v, new %v", oldR, newR)
+		return
+	}
+	// only react to status.phase change
+	// NOTE: currently changing reservation spec is not supported, and the new spec will be ignored.
+	if oldR.Status.Phase == newR.Status.Phase {
+		klog.V(5).InfoS("skip for reservation update since its status.phase is not change",
+			"reservation", reservation.GetReservationNamespacedName(newR))
+		return
+	}
+	klog.V(3).InfoS("got reservation phase update",
+		"reservation", reservation.GetReservationNamespacedName(newR), "old", oldR.Status.Phase, "new", newR.Status.Phase)
+
+	if oldR.Status.Phase == schedulingv1alpha1.ReservationAvailable &&
+		newR.Status.Phase == schedulingv1alpha1.ReservationExpired {
+		deleteReservation(sched, newR)
+	}
+}
+
+func deleteReservation(sched *scheduler.Scheduler, r *schedulingv1alpha1.Reservation) {
+	if r == nil || r.Spec.Template == nil {
+		klog.Errorf("failed to delete reservation for invalid reservation %v", r)
+		return
+	}
+
+	reservePod := reservation.NewReservePod(r)
+	// if reservation is scheduled, delete it from the schedule cache;
+	// otherwise, remove it from the scheduling queue.
+
+	if reservation.IsReservationScheduled(r) {
+		err := sched.SchedulerCache.RemovePod(reservePod)
+		if err != nil {
+			klog.Errorf("failed to remove reserve pod in scheduler cache, reservation %v, err: %v", reservation.GetReservationNamespacedName(r), err)
+		}
+		sched.SchedulingQueue.MoveAllToActiveOrBackoffQueue(assignedPodDelete, nil)
+	} else {
+		// in case the pod is just scheduled and deleted, also check if pod is cached
+		_, err := sched.SchedulerCache.GetPod(reservePod)
+		if err == nil {
+			klog.V(5).InfoS("reserve pod is just scheduled and deleted, remove it in cache",
+				"reservation", reservation.GetReservationNamespacedName(r))
+			err = sched.SchedulerCache.RemovePod(reservePod)
+			if err != nil {
+				klog.Errorf("failed to remove reserve pod in scheduler cache, reservation %v, err: %v", reservation.GetReservationNamespacedName(r), err)
+			}
+			sched.SchedulingQueue.MoveAllToActiveOrBackoffQueue(assignedPodDelete, nil)
+		}
+
+		err = sched.SchedulingQueue.Delete(reservePod)
+		if err != nil {
+			klog.Errorf("failed to delete reserve pod in scheduling queue, reservation %v, err: %v", reservation.GetReservationNamespacedName(r), err)
+		}
+		fwk, ok := sched.Profiles[reservePod.Spec.SchedulerName]
+		if ok {
+			fwk.RejectWaitingPod(reservePod.UID)
```

Comment:
According to the current implementation, there is no need to deal with the waiting state.

Reply:
fixed